azure-kit-starter/src/main/java/com/vaadin/azure/starter/ui/ClusterSupport.java

```diff
@@ -0,0 +1,171 @@
+        serviceInitEvent.addRequestHandler(this::handleRequest);
+    }
+
+    private boolean handleRequest(VaadinSession vaadinSession,
+            VaadinRequest vaadinRequest, VaadinResponse vaadinResponse) {
+        // Set the thread local instance
+        CurrentInstance.set(ClusterSupport.class, this);
+
+        vaadinSession.access(() -> {
+            // Always set the version cookie
+            Cookie currentVersionCookie = getCookieByName(
+                    CURRENT_VERSION_COOKIE);
+            if (currentVersionCookie == null
+                    || !currentVersionCookie.getValue().equals(appVersion)) {
+                currentVersionCookie = new Cookie(CURRENT_VERSION_COOKIE,
+                        appVersion);
+                currentVersionCookie.setHttpOnly(true);
+                vaadinResponse.addCookie(currentVersionCookie);
+            }
+
+            // Always check for the new version cookie
+            Cookie updateVersionCookie = getCookieByName(UPDATE_VERSION_COOKIE);
+            if (updateVersionCookie != null
+                    && !updateVersionCookie.getValue().isEmpty()
+                    && !currentVersionCookie.getValue()
+                            .equals(updateVersionCookie.getValue())) {
+                vaadinSession.getUIs().forEach(ui -> {
+                    if (ui.getChildren().anyMatch(
+                            child -> (child instanceof VersionNotificator))) {
+                        return;
+                    }
+                    VersionNotificator notificator = new VersionNotificator(
+                            appVersion, updateVersionCookie.getValue());
+                    notificator.addSwitchVersionEventListener(
+                            this::handleSwitchVersionEvent);
+                    // Show notificator
+                    ui.add(notificator);
+                });
+            }
+        });
+
+        // If the current and the new versions do not match notify the user
+        return false;
+    }
+
+    private void handleSwitchVersionEvent(
+            VersionNotificator.SwitchVersionEvent event) {
+        if (switchVersionListener != null) {
+            // Do nothing if switch version listener prevents switching
+            if (!switchVersionListener.nodeSwitch(VaadinRequest.getCurrent(),
+                    VaadinResponse.getCurrent())) {
+                return;
+            }
+
+            // Do application level clean-up before version switch
+            switchVersionListener.doAppCleanup();
+        }
+
+        // When the user clicks on the notificator remove the cluster
+        // and session cookies
+        removeCookie(STICKY_CLUSTER_COOKIE);
+        removeCookie(CURRENT_VERSION_COOKIE);
+        removeCookie(UPDATE_VERSION_COOKIE);
+
+        // Invalidate the session, Vaadin does not synchronizes it
+        // between clusters
+        VaadinRequest.getCurrent().getWrappedSession().invalidate();
+    }
```

Comment:
This method is still not tested. We can trigger the event on `VersionNotificator` programmatically, or make the method package protected and call it on a test with a mock event. Either way, we must verify cookies are correctly removed and the session is invalidated and app cleanup actually happens.